modules/basegl/glsl/raycasting/raycaster-template.frag

```diff
@@ -0,0 +1,117 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+// The special //? and //! comments are helpers for the visual studio shader validation
+
+//? #version 460
+
+#include "utils/structs.glsl"   //! #include "../../../opengl/glsl/utils/structs.glsl"
+#include "utils/sampler3d.glsl" //! #include "../../../opengl/glsl/utils/sampler3d.glsl"
+#include "utils/depth.glsl"     //! #include "../../../opengl/glsl/utils/depth.glsl"
+
+#pragma IVW_SHADER_SEGMENT_PLACEHOLDER_INCLUDE
+
+float calcStep(in float rayLength, in vec3 direction, in float samplingRate, in vec3 dimensions) {
+    float incr = min(rayLength, rayLength / (samplingRate * length(direction * dimensions)));
+    float samples = ceil(rayLength / incr);
+    return rayLength / samples;
+}
+
+vec3 calcCameraDir(in vec3 entryPoint, in vec3 exitPoint, in mat4 textureToWorld) {
+    return normalize(
+        (textureToWorld * vec4(entryPoint, 1.0) - textureToWorld * vec4(exitPoint, 1.0)).xyz);
+}
+
+
+layout(location = 0) out vec4 FragData0;
+layout(location = 1) out vec4 PickingData;
+
+uniform ImageParameters outportParameters;
+uniform float samplingRate = 2.0;
+
+//? uniform VolumeParameters volumeParameters;
+//? uniform CameraParameters camera;
+
+#pragma IVW_SHADER_SEGMENT_PLACEHOLDER_UNIFORM
+
+
+void main() {
+    vec4 result = vec4(0.0);  // The accumulated color along the ray;
+    vec4 picking = vec4(0.0); // The picking color of the ray
+    float rayDepth = -1.0;    // The ray depth value (0 to ray length, -1 mean "no" depth
+    float depth = 1.0;        // The image depth
+
+    vec2 texCoords = gl_FragCoord.xy * outportParameters.reciprocalDimensions;
+ 
+    // The setup is expected to define
+    // entryPoint, exitPoint, entryPointDepth, exitPointDepth, rayLength, rayDirection
```

Comment:
I did not want to declare them without setting them. Also this way you will get an error if the setup does not set them. If I would declare them it would still compile but not be correct...